### ISSUE TITLE
history pagebeforechange triggering handler + changePage twice in 1.1.0 RC1

### ISSUE THREAD

**Issue by raldred** [azicchetti/jquerymobile-router#31]
When jqm changes page it triggers the pagebeforechange event twice,
before 1.1.0 RC1 this was fine, I assume the router handled this and only called changePage and route handlers once

However it seems to being called 4 times now and triggering changePage twice and the handlers twice, when you navigation the history with the jqm back button or the browser buttons you get nasty double transition, first back and then forward to the same page.
It seems like the _jqmrouter_handled option seems to be set on the data passed.
However on the 3rd and 4th calls to pagebeforechange have the allowSamePageTransition set to true

Something else to note, it only happens when the page you change to in history has query parameters
eg. #section?id=11
pages without parameters _don't_ trigger the double transition.
eg. #page

Out of interested, commenting out these two lines:

```
      67: data.options.allowSamePageTransition=true;
      68: $.mobile.changePage( $(page), data.options );
```

fixes this specific issue, i suspect it might break something else though?

Also, switching back to 1.0.1 returns everything to how it was, handlers and changePage is only triggered once.

Appreciate your help.

---

I inserted a log call on line 41 of`jquery.mobile.router` and on the first line of the handler for the page I was testing with:

```
  console.log('<<pagebeforechange>>');
```

Log with 1.0.1

```
<<pagebeforechange>>  jquery.mobile.router.js:41
<<pagebeforechange>>  jquery.mobile.router.js:41
showSection Handler   jam.editor.actions.js:150
```

Log with 1.1.0-RC.1

```
<<pagebeforechange>>  jquery.mobile.router.js:41
<<pagebeforechange>>  jquery.mobile.router.js:41
showSection Handler   jam.editor.actions.js:150
<<pagebeforechange>>  jquery.mobile.router.js:41
<<pagebeforechange>>  jquery.mobile.router.js:41
showSection Handler   jam.editor.actions.js:150
```

**Comment by azicchetti** [comment]
Hi raldred,
thanks for digging into this problem.
As far as I can remember, even 1.0.1 triggers pagebeforechange twice and the router tries to discard the second call by using the property _jqmrouter_handled [in previous versions I was using a different approach that wasn't supporting explicit $.mobile.changePage's with a jquery object as the first argument].

I've done a couple of experiments (jquery 1.7.1 and jqm rc1) and everything seems consistent to me with the previous 1.0.1 version, so I'm probably missing something important (for instance, I haven't tried a programmatic $.mobile.changePage).
Would you mind sharing your example?

Thank you

**Comment by raldred** [comment]
Hi Andrea, thanks for getting back to me about this.
I've prepared a couple of examples, the same except the version of jqm, both using the current master of mobile router.

To replicate..
Click "Section 1"
Click "Chunk 1"
Click "Back"

In the 1.1.0 RC1 example clicking "Back" will cause the #section page to transition twice.

1.1.0 RC1
http://jsfiddle.net/robaldred/BJvHf/

1.0.1
http://jsfiddle.net/robaldred/ufnGy/

(Tested with Chrome Mac 17.0.963.65)

**Comment by azicchetti** [comment]
A quick followup until I have more time to dig into the problem.

It seems that the 1.1.0 version will support parameters in the hash part of the url natively (this is very good news!) and so the

``` javascript
if ( u.hash.indexOf("?") !== -1 ) {
...
}
```

branch in the router is no longer needed.
Removing it apparently fixes the problem.
It would be great if you had a little time to confirm it.

Thanks!

**Comment by raldred** [comment]
Hi again,
Interesting, that is pretty good news.

I've altered the router to remove the addon support for parameters.
the pagebeforechange now looks like....

```
$(document).bind("pagebeforechange", function( e, data ) {
  var toPage=( typeof data.toPage === "string" ) ? data.toPage : data.toPage.jqmData("url")||"";

  if ( data.options.hasOwnProperty("_jqmrouter_handled") ){ return; }
  data.options._jqmrouter_handled = true;

  var u = $.mobile.path.parseUrl( toPage );
  previousUrl=nextUrl;
  nextUrl=u;

  $.mobile.changePage(toPage, data.options);
  e.preventDefault();
});
```

It seems to work correctly, however I might need to do some further testing.

**Comment by azicchetti** [comment]
Uhm, try deleting the last two lines (I think the problem here was the changePage. Basically, that part of the router code is now in jquery mobile itself, so we need to remove it to avoid duplicates)

``` javascript
      $.mobile.changePage(toPage, data.options);
      e.preventDefault();
```

**Comment by raldred** [comment]
great stuff, that works aswell.
So i literally just have

```
$(document).bind("pagebeforechange", function( e, data ) {
  var toPage=( typeof data.toPage === "string" ) ? data.toPage : data.toPage.jqmData("url")||"";

  if ( data.options.hasOwnProperty("_jqmrouter_handled") ){ return; }
  data.options._jqmrouter_handled = true;

  var u = $.mobile.path.parseUrl( toPage );
  previousUrl=nextUrl;
  nextUrl=u;
});
```

**Comment by azicchetti** [comment]
yep, that should be it!
I'm taking a few days to properly examine jquery mobile sources before pushing a patch, so please use your own version in the meanwhile.

**Comment by raldred** [comment]
Sure thing, thanks for your help, really appreciate your work on the plugin, it's been a great addition to my app.

**Comment by raldred** [comment]
Hey, just found an issue with the modifications we made.
We need the following 3 lines back in:

```
  data.options.dataUrl = u.href;
  $.mobile.changePage(toPage, data.options);
  e.preventDefault();
```

Without these browsing to and from the same path with different params does not change the browser hash or the jquery history to reflect the new params.

**Comment by azicchetti** [comment]
You mean browsing from #page?foo=bar to #page?bar=foo ?

If we put back those lines we may have problems with ajax pages, actually I guess it would break them completely.

I think that with a simple check on the active page we may fix this bug without too much effort, but it would be better if it was fixed in jquery mobile itself, since it seems a problem in their newly added code to support parameters in the hash.

Do you mind making a super-simple test page without the router and submit the bug report (if it still applies!) to jquery mobile guys?
If they won't fix it, I'll hopefully make a patch before the final 1.1.0 is released

**Comment by raldred** [comment]
Ahh damn, it also breaks dialogs. I'll try but im not entirely sure what the issue with jqm's internals are, dont suppose you could point me to the section youve been looking at?

**Comment by raldred** [comment]
Ok, dirty, but for the time being this works for my particular application (Dialogs & Ajax pages also work fine).
Also, im using the jqm simpledialog2 plugin which works aswell now.

```
  if(u.href != "") {
    data.options.dataUrl = u.href;
    $.mobile.changePage(toPage, data.options);
    e.preventDefault();
  }
```

**Comment by azicchetti** [comment]
I'm supposing that the problem we're talking about is browsing from #page?foo=bar to #page?bar=foo, is it correct?

I think the problem with jqm is that they've added support for hash parameters in the url but they forgot to enable same page transitions.

Try replacing your #section page with this one:

```
<div id="section" data-role="page" data-add-back-btn="true">
    <div data-role="header"><h1>Section</h1></div>
    <div data-role="content">
        <ul data-role="listview">
            <li><a href="#section?id=2">Same page id=2</a></li>
            <li><a href="#section?id=3">Same page id=3</a></li>
            <li><a href="#section?id=4">Same page id=4</a></li>
        </ul>
    </div>
</div>
```

Don't load the jqm router.
You'll notice that clicking any of the "same page" links won't do anything (this is the jqm bug, I suppose).

Now, try loading the original jqm-router version (the unpatched one). Links are now working as expected.
jQuery mobile should handle this scenario natively

**Comment by azicchetti** [comment]
Hi,
I've pushed a fix to support jqm 1.1.0rc2.

It turns out that the 'native' support for hash parameters is a side effect of a jquery selector used in the loadPage() function of jQM.
Since I'm scared by this, I'm keeping the code we removed a couple of weeks ago in the router, because it feels more safe and sound. 

Hopefully the handler won't be called twice with the latest fix.
